### PR TITLE
uuid registry support and added uuid to patient_data

### PR DESCRIPTION
**Pull request by bradymiller**
Adds main mechanism for supporting/tracking uuid. Used the mechanism to add uuid support to patient_data table. Still need to test the upgrade mechanism.

Built from prior PRs:
https://github.com/openemr/openemr/pull/2360
https://github.com/openemr/openemr/pull/2574

Fixes issue:
https://github.com/openemr/openemr/issues/2277

Also fixed double escaping in BaseService class

Also fixes some php notices that were showing up on automated testing

### CODE REVIEW

File: src/Common/Uuid/UuidRegistry.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    public function createUuid()
+    {
+        $isUnique = false;
+        $i = 0;
+        while (!$isUnique) {
+            $i++;
+            if ($i > self::MAX_TRIES) {
+                // This error should never happen. If so, then the random generation of the
+                //  OS is compromised and no use continuing to run OpenEMR.
+                error_log("OpenEMR Error: API was unable to create a unique UUID");
```
Comment (bradymiller) [inline src/Common/Uuid/UuidRegistry.php:47]: message should be:
```
error_log("OpenEMR Error: Unable to create a unique UUID");
```

File: src/Common/Uuid/UuidRegistry.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        // Return the uuid
+        return $uuid;
+    }
+
+    // Generic function to create missing uuids in a sql table (table needs an `id` column to work)
+    public function createMissingUuids()
+    {
+        $resultSet = sqlStatementNoLog("SELECT id FROM " . $this->table_name . " WHERE uuid = ''");
+        while ($row = sqlFetchArray($resultSet)) {
+            $uuid = self::createUuid();
```
Comment (bradymiller) [inline src/Common/Uuid/UuidRegistry.php:82]:
```
$uuid = $this->createUuid();
```

File: src/Services/BaseService.php [hunk 3]
```diff
@@ -80,7 +80,7 @@ protected function buildInsertColumns($passed_in = array())
                 $value = "";
             }
             $keyset .= ($keyset) ? ", `$key` = ? " : "`$key` = ? ";
-            $bind[] = ($value === null || $value === false) ? '' : add_escape_custom($value);
+            $bind[] = ($value === null || $value === false) ? '' : $value;
```
Comment (bradymiller) [inline src/Services/BaseService.php:83]: these are binded, so don't need escaping

File: src/Services/BaseService.php [hunk 4]
```diff
@@ -119,7 +119,7 @@ protected function buildUpdateColumns($passed_in = array())
                 continue;
             }
             $keyset .= ($keyset) ? ", `$key` = ? " : "`$key` = ? ";
-            $bind[] = ($value == 'NULL') ? "" : add_escape_custom($value);
+            $bind[] = ($value == 'NULL') ? "" : $value;
```
Comment (bradymiller) [inline src/Services/BaseService.php:122]: these are binded, so don't need escaping

**Comment by bradymiller** [comment]
travis is angry :scream: 

image

**Comment by bradymiller** [comment]
feels good to beat travis :)

tenor (19)

**Comment by bradymiller** [comment]
@sjpadgett , @dixonwhitmire , @yashrajbothra , @ShubhamBansal1997 
This is currently testing well. Incorporated it into the patient_data table so every entry will have a unique uuid (also will be case for users that upgrade from prior openemr versions). Also straightforward to incorporate this into other tables as needed. Also won't be tough to add capability to collect/store for transactional stuff. Plan to bring in soon if no issues are brought up on review.

**Comment by dixonwhitmire** [comment]
Thanks @bradymiller ! When this is ready to go I'll merge it on in with my current feature branch, so that I can wire things up. Thanks!

**Comment by bradymiller** [comment]
@dixonwhitmire , This is in codebase now, so can do your magic :)

**Comment by dixonwhitmire** [comment]
Thanks @bradymiller ! I've merged in the latest from master. It looks like I have a few test cases that are failing, so I'll look into that and will keep 🚜 forward!

**Pull request merged**